### PR TITLE
ContainerBuilder::getClassList: optimized performance

### PR DESCRIPTION
**Pull request by matej21**

### CODE REVIEW

File: src/DI/ContainerBuilder.php [hunk 1]
```diff
@@ -259,14 +259,24 @@ public function findByTag($tag)
 	private function getClassList()
 	{
 		static $prev;
-		if ($this->classes !== FALSE && $prev !== serialize($this->definitions)) {
+		if ($this->classes !== FALSE && count($prev) !== count($this->definitions) && $prev !== $this->getDefinitionsInfo()) {
 			$this->prepareClassList();
-			$prev = serialize($this->definitions);
+			$prev = $this->getDefinitionsInfo();
```
Comment (JanTvrdik) [inline src/DI/ContainerBuilder.php]: You call `$this->getDefinitionsInfo` twice.

**Comment by dg** [comment]
What's the trick?

**Comment by matej21** [comment]
@dg the trick is that only class/factory/implement or autowire can affect the result of prepareClassList. e.g. setup or inject can't.

**Comment by JanTvrdik** [comment]
What if you add one definitions and remove another one?

**Comment by matej21** [comment]
@JanTvrdik good point. will fix..

**Comment by matej21** [comment]
oh, it is a bit wrong. wait a minute :)

**Comment by dg** [comment]
Where is the problem? Serialize takes a lot of time?

**Comment by matej21** [comment]
In fact, that count is totally wrong. But without that performance goes from ~2s to ~8s (but still better than >30s)

@dg serialize is one problem. The second (and probably more important) is that prepareClassList is called many times, even it is not necessary (e.g. when setup method was added)

File: src/DI/ContainerBuilder.php [hunk 2]
```diff
@@ -259,14 +259,24 @@ public function findByTag($tag)
 	private function getClassList()
 	{
 		static $prev;
-		if ($this->classes !== FALSE && $prev !== serialize($this->definitions)) {
+		if ($this->classes !== FALSE && count($prev) !== count($this->definitions) && $prev !== $this->getDefinitionsInfo()) {
```
Comment (castamir) [inline src/DI/ContainerBuilder.php]: You don't need to call `getDefinitionsInfo()` twice, do you?

**Comment by JanTvrdik** [comment]
My guess is that you need to track dependencies of each individual class in class list.

**Comment by castamir** [comment]
Got an idea, it's maybe a little offtopic, but still... is there a way, how to create a file watcher for e.g. PhpStorm? :D 

It might be interesting...

File: src/DI/ContainerBuilder.php (hunk 2 above)
Comment (matej21) [inline src/DI/ContainerBuilder.php]: I do, its result is affected by prepareClassList

**Comment by matej21** [comment]
rewritten

File: src/DI/ContainerBuilder.php [hunk 3]
```diff
@@ -890,4 +898,13 @@ public function getServiceName($arg)
 		return $service;
 	}
 
+
+	/**
+	 * @internal
+	 */
+	public function notifyDefinitionChanged()
```
Comment (JanTvrdik) [inline src/DI/ContainerBuilder.php]: that's ugly why not just use events

Comment (matej21) [inline src/DI/ContainerBuilder.php]: I was worried about a performance. But it is ok... 

updated, what do you think now?

**Comment by JanTvrdik** [comment]
But that's still slow, no?

File: src/DI/ContainerBuilder.php [hunk 4]
```diff
@@ -59,14 +62,23 @@ class ContainerBuilder
...
 		if (isset($this->definitions[$name])) {
 			throw new Nette\InvalidStateException("Service '$name' has already been added.");
 		}
-		return $this->definitions[$name] = $definition ?: new ServiceDefinition;
+		if (!$definition) {
+			$definition = new ServiceDefinition();
+		}
+		$definition->onChange[] = function ($service, $method) {
+			if (in_array($method, ['setClass', 'setFactory', 'setImplement', 'setAutowired'], TRUE)) {
+				$this->definitionsChanged = TRUE;
```
Comment (enumag) [inline src/DI/ContainerBuilder.php]: Definitions changed even with the other methods. Maybe rename it to sth like classListObsolete?

**Comment by matej21** [comment]
@JanTvrdik performance is great now. ~0.7s now vs ~0.5s without that commit which regenerates class list

File: src/DI/ContainerBuilder.php (hunk 4 above)
Comment (matej21) [inline src/DI/ContainerBuilder.php]: Not sure I understand...

**Comment by JanTvrdik** [comment]
Interesting. You still regenerate the class list more often that is necessary. But I guess that other cases are not nearly as important.

I was playing with [tracking dependencies per class](https://github.com/JanTvrdik/nette-di/commit/73a6ff34b44aaf4cedce4941ad5f30371b5b3ddc). But as it turned out, you can't really implement it this way without tracking calls to setImplement + setClass.

**Comment by matej21** [comment]
> Interesting. You still regenerate the class list more often that is necessary. 

for my container it is regenerated ~10 times, that is ok

File: src/DI/ContainerBuilder.php (hunk 4 above)
Comment (matej21) [inline src/DI/ContainerBuilder.php]: Oh, I understand now. Yes, that is better name :)

**Comment by dg** [comment]
It would be great to put together a sandbox for doing DI benchmarks. With a greater number of services and with some extensions.

**Comment by dg** [comment]
This is a nice solution, but unfortunately Statement is mutable too.

**Comment by dg** [comment]
It may be finished it two ways:
1) replace __FUNCTION__ with attribute name to eliminate differences between setSetup() and addSetup()
2) make it as simple as possible and track only what we need, ie. setClass, setFactory, setImplement, setAutowired and additionally setEntity for entities added via setFactory. Notifier can be simple callback instead of array of callbacks.

I more like option 2.

**Comment by matej21** [comment]
@dg how do you like the solution before this commit https://github.com/nette/di/pull/108/commits/9e680d94f62a5ffb17b006e0e0252646f6a821ef ? it was similar to your option 2

**Comment by dg** [comment]
Yes, it is. It only differs in passed callback vs passed builder. I am not sure what is better.

**Comment by dg** [comment]
Statement::setEntity is used extremely rare (two occurrences in Kdyby), probably best thing is to make it immutable in future and trigger warning in 2.4

**Comment by matej21** [comment]
> I am not sure what is better.

Me neither. But passing CB is more straightforward and it cannot be abused by a user.

> probably best thing is to make it immutable in future and trigger warning in 2.4

:+1: so can I ignore it? and just revert commit  9e680d9?

**Comment by dg** [comment]
Done 8211907ac01d5dc100fa01850c34d5a7873385e2

**Comment by matej21** [comment]
PR updated. What do you think?

File: src/DI/ServiceDefinition.php [hunk 5]
```diff
@@ -302,4 +316,14 @@ public function getInject()
 		return $this->getTag(Extensions\InjectExtension::TAG_INJECT);
 	}
 
+
+	/**
+	 * @param ContainerBuilder
+	 * @internal
+	 */
+	public function setContainerBuilder(ContainerBuilder $containerBuilder)
```
Comment (dg) [inline src/DI/ServiceDefinition.php]: Or optional constructor parameter?

**Comment by dg** [comment]
I think that unserialize(serialize()) will create a clone of ContainerBuilder too.

**Comment by dg** [comment]
Hey, I can add support for deep clone directly into ServiceDefinition, right?

**Comment by matej21** [comment]
good point, updated. But do we still need a deep clone since Statement is (will be) immutable?

**Comment by dg** [comment]
c0258ca0c204ca0d8d27d5ab66c64427d0263f8b

**Comment by matej21** [comment]
rebased & squashed

**Comment by dg** [comment]
What do you think about this https://github.com/dg/nette-di/commits/test - builder replaced with callback, so internal method is not needed

**Comment by matej21** [comment]
@dg I'm ok with that.

**Comment by dg** [comment]
$notifier = 'pi' is IMHO too tricky

**Comment by matej21** [comment]
it is. maybe assign an empty closure in the constructor?

**Comment by dg** [comment]
remove conditions + add constructor = 3 lines saved :-)

**Comment by dg** [comment]
On the other hand, π trick is unique and Nette will be the first web framework that use this function…

**Pull request closed without merging**